### PR TITLE
#337 improve generators algorithm when a rectangular reactive capabil…

### LINKED ISSUE #337: Use generator model without npoints diagram when reactive points describe a rectangle
this will reduce the number of file to written as txt, as well the number of variables

### PR DESCRIPTION
**Pull request by marcochiaramello**
…ity curve is specified

Closes #337 

**Checklist before requesting a review**

*use `'[x]'` to check the checkboxes, or submit the PR and then click the checkboxes*

- [x] unit tests and non regression tests were added (update of inputs, outputs, SA, algos)
- [x] main documentation was updated (update of input/output file)
- [ ] if this commit targets a release branch: the corresponding milestone was added in the ticket and in this PR

### CODE REVIEW

File: documentation/functionalDoc/functionalDoc.tex [hunk 1]
```diff
@@ -58,31 +58,39 @@ \subsection{Generators}
...
-\small{regulates the same node} & \multirow{-2}{*}{\small{true}} & \footnotesize{+ VRRemote model} \\
+\scriptsize{regulates the same node} & \multirow{-2}{*}{\scriptsize{false}} & \multirow{-2}{*}{\scriptsize{true}} & \scriptsize{+ VRRemote model} \\
+\rowcolor{gray!10}
+\scriptsize{Another generator}&  & & \scriptsize{GeneratorPQPropSignalN}  \\
+\rowcolor{gray!10}
+\scriptsize{regulates the same node} & \multirow{-2}{*}{\scriptsize{true}} & \multirow{-2}{*}{\scriptsize{-}} & \scriptsize{+ VRRemote model} \\
 \bottomrule
 \end{tabular}
-\caption{Generators modelling}
+\caption{Generators modelling. When a generator has a rectangular diagram, it means all the values of $Q_{min}$ are equal and all the values of $Q_{max}$ are equal}
```
Comment (rosiereflo) [inline documentation/functionalDoc/functionalDoc.tex:93]: Maybe better to put this above the table?

**Review by rosiereflo — CHANGES_REQUESTED** [review]

File: sources/Algo/src/GeneratorDefinitionAlgorithm.cpp [hunk 2]
```diff
@@ -153,5 +157,18 @@ GeneratorDefinitionAlgorithm::isTargetPValid(const inputs::Generator& generator)
...
 }
 
+bool
+notEqual(const DYN::GeneratorInterface::ReactiveCurvePoint& firstPoint, const DYN::GeneratorInterface::ReactiveCurvePoint& secondPoint) {
+  return DYN::doubleNotEquals(firstPoint.qmax, secondPoint.qmax) || DYN::doubleNotEquals(firstPoint.qmin, secondPoint.qmin);
+}
+
+bool
+GeneratorDefinitionAlgorithm::isDiagramRectangular(const inputs::Generator& generator) const {
+  if (std::adjacent_find(generator.points.begin(), generator.points.end(), notEqual) == generator.points.end()) {
```
[the remark this answers is not in the document]
Comment (marcochiaramello) [reply sources/Algo/src/GeneratorDefinitionAlgorithm.cpp]: thx!

File: documentation/functionalDoc/functionalDoc.tex (hunk 1 above)
Comment (marcochiaramello) [reply documentation/functionalDoc/functionalDoc.tex]: yes, I put it above and more explicitly, thx

File: sources/Algo/src/GeneratorDefinitionAlgorithm.cpp [hunk 3]
```diff
@@ -30,49 +30,50 @@ void
...
       } else {
         switch (nbOfRegulatingGenerators) {
         case dfl::inputs::NetworkManager::NbOfRegulating::ONE:
-          algoRes->isAtLeastOneGeneratorRegulating = true;
           if (generator.regulatedBusId == generator.connectedBusId) {
-            model = useInfiniteReactivelimits_ ? GeneratorDefinition::ModelType::SIGNALN : GeneratorDefinition::ModelType::DIAGRAM_PQ_SIGNALN;
+            if (useInfiniteReactivelimits_) {
+              model = ModelType::SIGNALN_INFINITE;
+            } else {
+              model = isDiagramRectangular(generator) ? ModelType::SIGNALN_RECTANGULAR : ModelType::DIAGRAM_PQ_SIGNALN;
```
Comment (cossartque) [inline sources/Algo/src/GeneratorDefinitionAlgorithm.cpp:55]: ok

**Review by cossartque — APPROVED** [review]

**Pull request merged**